dbms/src/Common/HashTable/TwoLevelStringHashTable.h

```diff
@@ -0,0 +1,217 @@
+    size_t hash(const Key & x) const
+    {
+        return const_cast<Self &>(*this).dispatch(x,
+            [&](const auto &, const auto &, size_t hash) { return hash; });
+    }
+
+    size_t operator()(const Key & x) const { return hash(x); }
+
+    /// NOTE Bad for hash tables with more than 2^32 cells.
+    static size_t getBucketFromHash(size_t hash_value) { return (hash_value >> (32 - BITS_FOR_BUCKET)) & MAX_BUCKET; }
+
+public:
+    using key_type = typename Impl::key_type;
+    using value_type = typename Impl::value_type;
+    using LookupResult = typename Impl::LookupResult;
+
+    Impl impls[NUM_BUCKETS];
+
+    TwoLevelStringHashTable() {}
+
+    template <typename Source>
+    TwoLevelStringHashTable(const Source & src)
+    {
+        if (src.m0.hasZero())
+            impls[0].m0.setHasZero(*src.m0.zeroValue());
+
+        for (auto & v : src.m1)
+        {
+            size_t hash_value = v.getHash(src.m1);
+            size_t buck = getBucketFromHash(hash_value);
+            impls[buck].m1.insertUniqueNonZero(&v, hash_value);
+        }
+        for (auto & v : src.m2)
+        {
+            size_t hash_value = v.getHash(src.m2);
+            size_t buck = getBucketFromHash(hash_value);
+            impls[buck].m2.insertUniqueNonZero(&v, hash_value);
+        }
+        for (auto & v : src.m3)
+        {
+            size_t hash_value = v.getHash(src.m3);
+            size_t buck = getBucketFromHash(hash_value);
+            impls[buck].m3.insertUniqueNonZero(&v, hash_value);
+        }
+        for (auto & v : src.ms)
+        {
+            size_t hash_value = v.getHash(src.ms);
+            size_t buck = getBucketFromHash(hash_value);
+            impls[buck].ms.insertUniqueNonZero(&v, hash_value);
+        }
+    }
+
+    // Dispatch is written in a way that maximizes the performance:
+    // 1. Always memcpy 8 times bytes
+    // 2. Use switch case extension to generate fast dispatching table
+    // 3. Combine hash computation along with bucket computation and key loading
+    // 4. Funcs are named callables that can be force_inlined
+    // NOTE: It relies on Little Endianness and SSE4.2
+    template <typename Func, typename KeyHolder>
+    decltype(auto) ALWAYS_INLINE dispatch(KeyHolder && key_holder, Func && func)
+    {
+        static constexpr StringKey0 key0{};
+        const StringRef & x = keyHolderGetKey(key_holder);
+        size_t sz = x.size;
+        const char * p = x.data;
+        // pending bits that needs to be shifted out
+        char s = (-sz & 7) * 8;
+        size_t res = -1ULL;
+        size_t buck;
+        union
+        {
+            StringKey8 k8;
+            StringKey16 k16;
+            StringKey24 k24;
+            UInt64 n[3];
+        };
+        StringHashTableHash hash;
+        switch (sz)
+        {
+            case 0:
+                keyHolderDiscardKey(key_holder);
```

Comment:
hmm, we don't persist yet, why discard?

Reply:
emplace() guarantees that it will either persist or discard the key. It doesn't pass the zero key_holder forward, and instead passes the empty `key0`, so it doesn't need the key from key_holder and should discard it here.

Actually I'm not sure when leaving it out would have any visible consequences  -- discard is only defined for SerializedKeyHolder, and if it is zero, it's going to have zero size, and the discard would be a noop.

Reply:
For uses other than emplace, it should always be the NoopKeyHolder, for which this call does nothing. I noticed that the signature of FindCallable::operator () is somewhat misleading now, because it accepts a template KeyHolder. I'll change it to only accept the NoopKeyHolder.

Reply:
Oops, we don't have the noop key holder anymore. It's represented by just the plain key now.

Reply:
I just realized I've answered the question about OnExistingKey and OnNewKey to you. Heh, now it confuses me. Those names do convey more than they should.

Reply:
Well, I think the persist/discard is better that onNew/onExisting, but maybe still bad indeed. Don't know how to rename it. Basically, it is a way for a hash table to say that it either needs a persistent version of this key, or it doesn't. Something like `makeKeyPersistent`/`dontMakeKeyPersistent`? This also may be bad, because for SerializedKeyHolder, `discardKey` actually destroys it, and you can't use it anymore.